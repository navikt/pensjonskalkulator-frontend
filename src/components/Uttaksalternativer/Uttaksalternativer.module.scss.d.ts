import globalClassNames from '../../style.d'
declare const classNames: typeof globalClassNames & {
  readonly uttaksalternativer: 'uttaksalternativer'
  readonly uttaksalternativerParagraph: 'uttaksalternativerParagraph'
  readonly uttaksalternativerChart: 'uttaksalternativerChart'
}
export = classNames
