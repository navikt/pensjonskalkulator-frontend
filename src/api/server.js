// const express = require('express')
// const liveness = require('../src/__mocks__/data/liveness.json')
// const pensjonsberegning = require('../src/__mocks__/data/pensjonsberegning.json')

// const app = express()
// const port = 8088

// app.get('/api/status', (request, response) => {
//   response.send(liveness)
// })

// app.get('/api/pensjonsberegning', async (request, response) => {
//   response.send(pensjonsberegning)
// })

// app.listen(process.env.PORT || port, () => {
//   console.info('proxy server is running on port', process.env.PORT || port)
// })

// import { rest } from 'msw'
// import { setupServer } from 'msw/node'

import { setupServer } from 'msw/node'
import { handlers } from './handlers.js'

export const server = setupServer(...handlers)
server.listen({ onUnhandledRequest: 'error' })
