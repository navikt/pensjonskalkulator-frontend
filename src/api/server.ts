import { setupServer } from 'msw/node'
import { getHandlers } from './handlers.js'
import { rest } from 'msw'

export const API_TARGET = 'http://localhost:8088'
export const API_PATH = '/pensjon/kalkulator/api'

const handlers = getHandlers(`${API_TARGET}${API_PATH}`)
export const server = setupServer(...handlers)

server.listen({ onUnhandledRequest: 'error' })

type MockResponseOptions = {
  status?: number
  json?: any
}

export const mockResponse = (
  path: string,
  options: MockResponseOptions = {
    status: 200,
    json: 'OK',
  }
) => {
  server.use(
    rest.get(`${API_TARGET}${API_PATH}${path}`, (req, res, ctx) => {
      return res(ctx.status(options.status ?? 200), ctx.json(options.json))
    })
  )
}

export const mockErrorResponse = (
  path: string,
  options: MockResponseOptions = {
    status: 500,
    json: "Beep boop I'm an error!",
  }
) => {
  mockResponse(path, {
    ...options,
    status: options.status ?? 500,
  })
}
