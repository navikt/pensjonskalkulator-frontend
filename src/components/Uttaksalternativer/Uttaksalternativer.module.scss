@import '../../scss/variables.scss';

.uttaksalternativer {
  background-color: var(--a-white);
  display: flex;
  flex-direction: column;
  gap: var(--a-spacing-4);
  padding: var(--a-spacing-6);
  text-align: start;
  margin-bottom: var(--a-spacing-10);

  @media (min-width: $a-breakpoint-sm) {
    padding: var(--a-spacing-10) var(--a-spacing-10);
  }

  &Paragraph {
    max-width: 75ch;
  }
}

.uttaksalternativerChart {
  padding: var(--a-spacing-2) 0 var(--a-spacing-4);
  width: 100%;

  @media (min-width: $a-breakpoint-md) {
    width: 20rem;
  }
}
