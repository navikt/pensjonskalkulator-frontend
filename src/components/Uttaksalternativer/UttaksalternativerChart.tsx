import { BarChart, ChartData } from '../BarChart'

const mapToChartData = (
  lønn: number,
  beregning: Pensjonsberegning[]
): ChartData[] => {
  const data = [...beregning]
    .sort((a, b) => a.alder - b.alder)
    .map(({ pensjonsbeloep, alder }) => ({
      value: pensjonsbeloep,
      label: `${alder} år`,
    }))

  data.unshift({ value: lønn, label: 'Lønn' })

  return data
}

interface Props {
  lønn: number
  beregning: Pensjonsberegning[]
}

export function UttaksalternativerChart({ lønn, beregning }: Props) {
  const data = mapToChartData(lønn, beregning)

  return <BarChart data={data} />
}
