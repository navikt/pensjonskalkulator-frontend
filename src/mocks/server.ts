import { rest } from 'msw'
import { setupServer } from 'msw/node'

import { API_BASEURL } from '@/paths'

import { getHandlers } from './handlers'

const handlers = getHandlers(API_BASEURL)
export const server = setupServer(...handlers)

type MockResponseOptions = {
  status?: number
  json?: ReturnType<typeof JSON.parse>
  text?: string
  method?: 'post' | 'get'
  baseUrl?: string
}

const defaultResponseOptions: Required<MockResponseOptions> = {
  status: 200,
  json: { data: 'OK' },
  method: 'get',
  baseUrl: `${API_BASEURL}`,
  text: '',
}

export const mockResponse = (
  path: string,
  inputOptions: MockResponseOptions = {}
) => {
  const options = { ...defaultResponseOptions, ...inputOptions }
  const parsedPath = `${options.baseUrl}${path}`

  server.use(
    rest[options.method](parsedPath, (_req, res, ctx) => {
      return res(
        ctx.status(options.status),
        options.text ? ctx.text(options.text) : ctx.json(options.json)
      )
    })
  )
}

const defaultErrorResponseOptions: Required<MockResponseOptions> = {
  status: 500,
  json: { data: "Beep boop I'm an error!" },
  text: '',
  method: 'get',
  baseUrl: `${API_BASEURL}`,
}

export const mockErrorResponse = (
  path: string,
  inputOptions: MockResponseOptions = {}
) => {
  const options = {
    ...defaultErrorResponseOptions,
    ...inputOptions,
  }
  mockResponse(path, options)
}
