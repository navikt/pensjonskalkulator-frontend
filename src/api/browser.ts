import { setupWorker } from 'msw'
import { getHandlers } from './handlers'
import { apiPath } from './server'

const handlers = getHandlers(apiPath)
export const worker = setupWorker(...handlers)
