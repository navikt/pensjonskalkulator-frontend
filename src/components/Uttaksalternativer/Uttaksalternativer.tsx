import React from 'react'

import {
  Alert,
  BodyLong,
  Button,
  Heading,
  Link,
  Loader,
} from '@navikt/ds-react'

import { useGetPensjonsberegningQuery } from '../../state/api/apiSlice'
import { isPensjonsberegning } from '../../state/api/typeguards'
import { formatAsDecimal } from '../../utils/currency'

import { UttaksalternativerChart } from './UttaksalternativerChart'

import styles from './Uttaksalternativer.module.scss'

const useInntekt = () => {
  return 678_000
}

export function Uttaksalternativer() {
  const { data, isLoading, isError } = useGetPensjonsberegningQuery()
  const inntekt = useInntekt()

  if (isLoading && !data) {
    return (
      <Loader
        className={styles.loader}
        data-testid="loader"
        size="3xlarge"
        title="Henter pensjonsberegning"
      />
    )
  }

  // TODO: Her bør det være unødvendig å sjekke om data er pensjonsberegning siden det burde være fanget opp av isError.
  //  Hvordan kan man få RTK Query / Typescript til så skjønne det? Har ikke redux-teamet fått til type narrowing automatisk?
  if (isError || !isPensjonsberegning(data)) {
    return (
      <Alert variant="error">
        <Heading spacing size="small" level="3">
          Vi klarte ikke å kalkulere pensjonen din. Prøv igjen senere.
        </Heading>
      </Alert>
    )
  }

  return (
    <section className={styles.uttaksalternativer}>
      <Heading size="small" level="3">
        Alternativer for når du kan ta ut
      </Heading>
      <BodyLong className={styles.uttaksalternativerParagraph}>
        Hvis du fortsetter å ha en inntekt på{' '}
        <strong>{formatAsDecimal(inntekt)} kr</strong> kan du tidligst gå av med
        alderspensjon når du blir <strong>{data[0].alder} år</strong>. Hvis du
        går av senere, får du høyere pensjon.
      </BodyLong>
      <section
        aria-label="Pensjonsberegning"
        className={styles.uttaksalternativerChart}
      >
        <UttaksalternativerChart lønn={inntekt} beregning={data} />
        <Button variant="secondary">Sjekk hele pensjonen din</Button>
      </section>
      <Link href="#">Om hvordan vi beregner din pensjon</Link>
    </section>
  )
}
