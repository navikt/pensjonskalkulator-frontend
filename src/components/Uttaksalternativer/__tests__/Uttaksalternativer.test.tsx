import { describe, expect, it } from 'vitest'

import { mockErrorResponse, mockResponse } from '../../../api/server'
import {
  render,
  screen,
  swallowErrorsAsync,
  waitFor,
} from '../../../test-utils'
import { Uttaksalternativer } from '../Uttaksalternativer'

describe('Uttaksalternativer', () => {
  it('viser loading og deretter pensjonsberegning hentet fra backend', async () => {
    const result = render(<Uttaksalternativer />)
    expect(screen.getByTestId('loader')).toBeVisible()

    await waitFor(() => {
      expect(screen.getByRole('heading', { level: 2 })).toHaveTextContent(
        'Når kan du ta ut alderspensjon?'
      )
      expect(
        screen.getByText('Hvis du fortsetter å ha en inntekt på', {
          exact: false,
        })
      ).toBeVisible()
      expect(screen.getByRole('link')).toHaveTextContent(
        'Om hvordan vi beregner din pensjon'
      )
      expect(screen.queryByTestId('loader')).not.toBeInTheDocument()
      expect(result.asFragment()).toMatchSnapshot()
    })
  })

  it('viser feilmelding om henting av pensjonberegning feiler', async () => {
    mockErrorResponse('/pensjonsberegning')

    const result = render(<Uttaksalternativer />)

    await waitFor(() => {
      expect(
        screen.getByText(
          'Vi klarte ikke å kalkulere pensjonen din. Prøv igjen senere.'
        )
      ).toBeVisible()
      expect(result.asFragment()).toMatchSnapshot()
    })
  })

  it('viser feilmelding om pensjonsberegning er på ugyldig format', async () => {
    const invalidData = {
      alder: 67,
      pensjonsaar: null,
    } as unknown as Pensjonsberegning
    mockResponse('/pensjonsberegning', { json: [invalidData] })

    render(<Uttaksalternativer />)

    await swallowErrorsAsync(async () => {
      await waitFor(() => {
        expect(
          screen.getByText(
            'Vi klarte ikke å kalkulere pensjonen din. Prøv igjen senere.'
          )
        ).toBeVisible()
      })
    })
  })
})
