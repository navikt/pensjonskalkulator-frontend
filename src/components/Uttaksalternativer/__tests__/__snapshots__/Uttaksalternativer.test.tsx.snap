// Vitest Snapshot v1

exports[`Uttaksalternativer > viser feilmelding om henting av pensjonberegning feiler 1`] = `
<DocumentFragment>
  <div
    class="navds-alert navds-alert--error navds-alert--medium"
  >
    <svg
      aria-labelledby="title-r2"
      class="navds-alert__icon"
      fill="none"
      focusable="false"
      height="1em"
      role="img"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title
        id="title-r2"
      >
        Feil
      </title>
      <path
        clip-rule="evenodd"
        d="M0 11.976C.013 5.372 5.395 0 11.999 0 18.639.013 24.012 5.406 24 12.022 23.986 18.627 18.603 24 12 24h-.023a11.924 11.924 0 0 1-8.48-3.531A11.923 11.923 0 0 1 0 11.976Zm17-3.547L15.571 7 12 10.572 8.429 7 7 8.429 10.572 12 7 15.571 8.429 17 12 13.428 15.571 17 17 15.571 13.428 12 17 8.429Z"
        fill="#C30000"
        fill-rule="evenodd"
      />
    </svg>
    <div
      class="navds-alert__wrapper navds-body-long"
    >
      <h3
        class="navds-heading navds-heading--small navds-typo--spacing"
      >
        Vi klarte ikke å kalkulere pensjonen din. Prøv igjen senere.
      </h3>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Uttaksalternativer > viser loading og deretter pensjonsberegning hentet fra backend 1`] = `
<DocumentFragment>
  <section
    class="_uttaksalternativer_2bdd5e"
  >
    <h3
      class="navds-heading navds-heading--medium"
    >
      Alternativer for når du kan ta ut
    </h3>
    <p
      class="_uttaksalternativerParagraph_2bdd5e navds-body-long"
    >
      Hvis du fortsetter å ha en inntekt på 
      <strong>
        678 000 kr
      </strong>
       kan du tidligst gå av med alderspensjon når du blir 
      <strong>
        67 år
      </strong>
      . Hvis du går av senere, får du høyere pensjon.
    </p>
    <section
      aria-label="Pensjonsberegning"
      class="_uttaksalternativerChart_2bdd5e"
    >
      <table
        class="_barchart_95eb26"
      >
        <thead>
          <tr
            class="_flexContainer_95eb26"
          >
            <th
              class="_barchartCaption_95eb26"
            >
              Lønn
            </th>
            <th
              class="_barchartCaption_95eb26"
            >
              67 år
            </th>
            <th
              class="_barchartCaption_95eb26"
            >
              68 år
            </th>
            <th
              class="_barchartCaption_95eb26"
            >
              69 år
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="_flexContainer_95eb26"
          >
            <td
              class="_barchartBar_95eb26"
              style="--barheight: 128px;"
            >
              <p
                class="navds-body-short"
              >
                678 000
              </p>
              <figure
                class="_barchartFigure_95eb26 _barchartFigure_isFirst_95eb26"
              />
            </td>
            <td
              class="_barchartBar_95eb26"
              style="--barheight: 56.63716814159292px;"
            >
              <p
                class="navds-body-short"
              >
                300 000
              </p>
              <figure
                class="_barchartFigure_95eb26"
              />
            </td>
            <td
              class="_barchartBar_95eb26"
              style="--barheight: 66.07669616519173px;"
            >
              <p
                class="navds-body-short"
              >
                350 000
              </p>
              <figure
                class="_barchartFigure_95eb26"
              />
            </td>
            <td
              class="_barchartBar_95eb26"
              style="--barheight: 75.51622418879056px;"
            >
              <p
                class="navds-body-short"
              >
                400 000
              </p>
              <figure
                class="_barchartFigure_95eb26"
              />
            </td>
          </tr>
        </tbody>
      </table>
      <button
        class="navds-button navds-button--secondary navds-button--medium"
      >
        <span
          aria-live="polite"
          class="navds-label"
        >
          Sjekk hele pensjonen din
        </span>
      </button>
    </section>
    <a
      class="navds-link"
      href="#"
    >
      Om hvordan vi beregner din pensjon
    </a>
  </section>
</DocumentFragment>
`;
